File: autocontext/src/autocontext/analytics/progress_report.py
"""Run progress curves, milestones, pass@k, and branch lineage reports."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal, Self

from pydantic import BaseModel, ConfigDict, Field

ProgressMilestoneName = Literal[
    "first_valid_candidate",
    "first_passing_verifier",
    "first_advancement",
    "threshold_success",
]

MILESTONE_NAMES: tuple[ProgressMilestoneName, ...] = (
    "first_valid_candidate",
    "first_passing_verifier",
    "first_advancement",
    "threshold_success",
)


class _StrictModel(BaseModel):
    model_config = ConfigDict(extra="forbid", strict=True)

    def to_dict(self) -> dict[str, Any]:
        return self.model_dump(mode="json")

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> Self:
        return cls.model_validate(data)


class ProgressPoint(_StrictModel):
    """One scored candidate on the best-score-over-time curve."""

    event_id: str = Field(min_length=1)
    timestamp: str = Field(min_length=1)
    elapsed_seconds: float
    generation_index: int | None = None
    hypothesis_node_id: str | None = None
    candidate_id: str | None = None
    score: float
    best_score: float
    improved: bool
    milestone_names: list[ProgressMilestoneName] = Field(default_factory=list)


class MilestoneTiming(_StrictModel):
    """Time to a named operator-facing progress milestone."""

    name: ProgressMilestoneName
    reached: bool
    event_id: str | None = None
    timestamp: str | None = None
    elapsed_seconds: float | None = None
    generation_index: int | None = None
    hypothesis_node_id: str | None = None
    score: float | None = None


class PassAtKSummary(_StrictModel):
    """Observed pass@k / best-of-k summary for the first k candidates."""

    k: int
    trials_considered: int
    successes: int
    passed: bool
    best_score: float | None
    threshold: float


class BranchLineageEdge(_StrictModel):
    """Parent/child hypothesis edge included in the inspection artifact."""

    parent_hypothesis_node_id: str = Field(min_length=1)
    child_hypothesis_node_id: str = Field(min_length=1)
    event_id: str = Field(min_length=1)
    generation_index: int | None = None


class RunProgressReport(_StrictModel):
    """Durable progress report shared by Python and TypeScript."""

    schema_version: Literal[1] = 1
    run_id: str = Field(min_length=1)
    generated_at: str = Field(min_length=1)
    threshold: float
    progress_points: list[ProgressPoint]
    milestones: list[MilestoneTiming]
    pass_at_k: list[PassAtKSummary]
    branch_lineage: list[BranchLineageEdge]

    def to_markdown(self) -> str:
        best_score = max((point.best_score for point in self.progress_points), default=None)
        pass_lines = [
            f"- pass@{summary.k}: {'pass' if summary.passed else 'miss'} "
            f"({summary.successes}/{summary.trials_considered}, best={summary.best_score})"
            for summary in self.pass_at_k
        ]
        milestone_lines = [
            f"- {milestone.name}: {milestone.elapsed_seconds:.3f}s"
            for milestone in self.milestones
            if milestone.reached and milestone.elapsed_seconds is not None
        ]
        parts = [
            f"# Progress Report: {self.run_id}",
            f"- Best score: {best_score}",
            f"- Threshold: {self.threshold}",
            "",
            "## Milestones",
            *(milestone_lines or ["- None reached"]),
            "",
            "## Pass@k",
            *(pass_lines or ["- No trials"]),
            "",
        ]
        return "\n".join(parts)


def build_run_progress_report(
    *,
    run_id: str,
    events: list[dict[str, Any]],
    threshold: float,
    pass_at_k_values: list[int] | None = None,
    generated_at: str | None = None,
) -> RunProgressReport:
    """Build a progress report from tree-search or campaign events."""

    ordered = sorted((_normalize_event(event) for event in events), key=_event_sort_key)
    start = _start_time(ordered, generated_at)
    scored = [event for event in ordered if _score(event) is not None]
    milestones_by_name = _milestones(ordered, scored, threshold, start)

    return RunProgressReport(
        run_id=run_id,
        generated_at=generated_at or datetime.now().astimezone().isoformat(),
        threshold=threshold,
        progress_points=_progress_points(scored, milestones_by_name, start),
        milestones=[milestones_by_name[name] for name in MILESTONE_NAMES],
        pass_at_k=_pass_at_k(scored, threshold, pass_at_k_values if pass_at_k_values is not None else [1, 5, 10]),
        branch_lineage=_branch_lineage(ordered),
    )


def progress_report_reference(report: RunProgressReport) -> dict[str, Any]:
    """Small inspection-safe pointer to a full progress report."""

    best_score = max((point.best_score for point in report.progress_points), default=None)
    return {
        "run_id": report.run_id,
        "threshold": report.threshold,
        "best_score": best_score,
        "milestones_reached": sum(1 for milestone in report.milestones if milestone.reached),
        "pass_at_k": [summary.model_dump(mode="json") for summary in report.pass_at_k],
    }


def _normalize_event(event: dict[str, Any]) -> dict[str, Any]:
    payload = event.get("payload")
    if not isinstance(payload, dict):
        return event
    event_id = event.get("event_id") or (f"seq-{event.get('seq')}" if event.get("seq") is not None else "")
    return {
        "event_id": event_id,
        "event_type": event.get("event"),
        "timestamp": event.get("ts") or event.get("timestamp"),
        "generation_index": payload.get("generation_index") or payload.get("generation"),
        "hypothesis_node_id": payload.get("hypothesis_node_id") or payload.get("node_id") or payload.get("child_id"),
        "parent_hypothesis_node_id": payload.get("parent_hypothesis_node_id") or payload.get("parent_id"),
        "candidate_id": payload.get("candidate_id") or payload.get("match_index"),
        "score": payload.get("score") if payload.get("score") is not None else payload.get("best_score"),
        "verifier_passed": payload.get("verifier_passed"),
        "gate_decision": payload.get("gate_decision") or payload.get("decision"),
    }


def _event_sort_key(event: dict[str, Any]) -> tuple[str, str]:
    return (_string(event.get("timestamp")), _string(event.get("event_id")))


def _start_time(events: list[dict[str, Any]], generated_at: str | None) -> datetime:
    timestamps = [_parse_time(_string(event.get("timestamp"))) for event in events if event.get("timestamp")]
    if timestamps:
        return min(timestamps)
    if generated_at:
        return _parse_time(generated_at)
    return datetime.now().astimezone()


def _milestones(
    events: list[dict[str, Any]],
    scored: list[dict[str, Any]],
    threshold: float,
    start: datetime,
) -> dict[ProgressMilestoneName, MilestoneTiming]:
    return {
        "first_valid_candidate": _milestone("first_valid_candidate", _first(scored), start),
        "first_passing_verifier": _milestone(
            "first_passing_verifier",
            _first(event for event in scored if event.get("verifier_passed") is True),
            start,
        ),
        "first_advancement": _milestone(
            "first_advancement",
            _first(event for event in events if _string(event.get("gate_decision") or event.get("decision")) == "advance"),
            start,
        ),
        "threshold_success": _milestone(
            "threshold_success",
            _first(event for event in scored if (_score(event) or 0.0) >= threshold),
            start,
        ),
    }


def _milestone(name: ProgressMilestoneName, event: dict[str, Any] | None, start: datetime) -> MilestoneTiming:
    if event is None:
        return MilestoneTiming(name=name, reached=False)
    timestamp = _string(event.get("timestamp"))
    return MilestoneTiming(
        name=name,
        reached=True,
        event_id=_string(event.get("event_id")),
        timestamp=timestamp,
        elapsed_seconds=_elapsed_seconds(timestamp, start),
        generation_index=_int_or_none(event.get("generation_index")),
        hypothesis_node_id=_str_or_none(event.get("hypothesis_node_id")),
        score=_score(event),
    )


def _progress_points(
    scored: list[dict[str, Any]],
    milestones_by_name: dict[ProgressMilestoneName, MilestoneTiming],
    start: datetime,
) -> list[ProgressPoint]:
    point_milestones: dict[str, list[ProgressMilestoneName]] = {}
    for milestone in milestones_by_name.values():
        if milestone.event_id:
            point_milestones.setdefault(milestone.event_id, []).append(milestone.name)

    points: list[ProgressPoint] = []
    best = float("-inf")
    for event in scored:
        score = _score(event)
        if score is None:
            continue
        improved = score > best
        if improved:
            best = score
        timestamp = _string(event.get("timestamp"))
        points.append(
            ProgressPoint(
                event_id=_string(event.get("event_id")),
                timestamp=timestamp,
                elapsed_seconds=_elapsed_seconds(timestamp, start),
                generation_index=_int_or_none(event.get("generation_index")),
                hypothesis_node_id=_str_or_none(event.get("hypothesis_node_id")),
                candidate_id=_str_or_none(event.get("candidate_id")),
                score=score,
                best_score=best,
                improved=improved,
                milestone_names=point_milestones.get(_string(event.get("event_id")), []),
            )
        )
    return points


def _pass_at_k(scored: list[dict[str, Any]], threshold: float, values: list[int]) -> list[PassAtKSummary]:
    scores = [_score(event) for event in scored]
    numeric_scores = [score for score in scores if score is not None]
    summaries: list[PassAtKSummary] = []
    for k in values:
        if k <= 0:
            continue
        window = numeric_scores[:k]
        successes = sum(1 for score in window if score >= threshold)
        summaries.append(
            PassAtKSummary(
                k=k,
                trials_considered=len(window),
                successes=successes,
                passed=successes > 0,
                best_score=max(window) if window else None,
                threshold=threshold,
            )
        )
    return summaries


def _branch_lineage(events: list[dict[str, Any]]) -> list[BranchLineageEdge]:
    seen: set[tuple[str, str]] = set()
    edges: list[BranchLineageEdge] = []
    for event in events:
        parent = _str_or_none(event.get("parent_hypothesis_node_id"))
        child = _str_or_none(event.get("hypothesis_node_id"))
        if not parent or not child or parent == child or (parent, child) in seen:
            continue
        seen.add((parent, child))
        edges.append(
            BranchLineageEdge(
                parent_hypothesis_node_id=parent,
                child_hypothesis_node_id=child,
                event_id=_string(event.get("event_id")),
                generation_index=_int_or_none(event.get("generation_index")),
            )
        )
    return edges


def _first(events: Any) -> dict[str, Any] | None:
    return next(iter(events), None)


def _score(event: dict[str, Any]) -> float | None:
    value = event.get("score")
    return float(value) if isinstance(value, int | float) and not isinstance(value, bool) else None


def _elapsed_seconds(timestamp: str, start: datetime) -> float:
    return (_parse_time(timestamp) - start).total_seconds()


def _parse_time(value: str) -> datetime:
    return datetime.fromisoformat(value.replace("Z", "+00:00"))


def _string(value: Any) -> str:
    return value if isinstance(value, str) else ""


def _str_or_none(value: Any) -> str | None:
    return value if isinstance(value, str) and value else None


def _int_or_none(value: Any) -> int | None:
    return value if isinstance(value, int) and not isinstance(value, bool) else None


__all__ = [
    "BranchLineageEdge",
    "MILESTONE_NAMES",
    "MilestoneTiming",
    "PassAtKSummary",
    "ProgressMilestoneName",
    "ProgressPoint",
    "RunProgressReport",
    "build_run_progress_report",
    "progress_report_reference",
]

File: ts/src/analytics/progress-report.ts
export const PROGRESS_MILESTONE_NAMES = [
  "first_valid_candidate",
  "first_passing_verifier",
  "first_advancement",
  "threshold_success",
] as const;

export type ProgressMilestoneName = (typeof PROGRESS_MILESTONE_NAMES)[number];

export interface RunProgressEvent {
  event_id: string;
  event_type: string;
  timestamp: string;
  generation_index?: number;
  hypothesis_node_id?: string;
  parent_hypothesis_node_id?: string;
  candidate_id?: string;
  score?: number;
  verifier_passed?: boolean;
  gate_decision?: string;
  decision?: string;
}

export interface RunProgressEventStreamRow {
  event?: string;
  ts?: string;
  seq?: number;
  payload?: Record<string, unknown>;
}

export type RunProgressEventInput = RunProgressEvent | RunProgressEventStreamRow;

export interface ProgressPoint {
  event_id: string;
  timestamp: string;
  elapsed_seconds: number;
  generation_index: number | null;
  hypothesis_node_id: string | null;
  candidate_id: string | null;
  score: number;
  best_score: number;
  improved: boolean;
  milestone_names: ProgressMilestoneName[];
}

export interface MilestoneTiming {
  name: ProgressMilestoneName;
  reached: boolean;
  event_id: string | null;
  timestamp: string | null;
  elapsed_seconds: number | null;
  generation_index: number | null;
  hypothesis_node_id: string | null;
  score: number | null;
}

export interface PassAtKSummary {
  k: number;
  trials_considered: number;
  successes: number;
  passed: boolean;
  best_score: number | null;
  threshold: number;
}

export interface BranchLineageEdge {
  parent_hypothesis_node_id: string;
  child_hypothesis_node_id: string;
  event_id: string;
  generation_index: number | null;
}

export interface RunProgressReport {
  schema_version: 1;
  run_id: string;
  generated_at: string;
  threshold: number;
  progress_points: ProgressPoint[];
  milestones: MilestoneTiming[];
  pass_at_k: PassAtKSummary[];
  branch_lineage: BranchLineageEdge[];
}

export interface BuildRunProgressReportInput {
  runId: string;
  events: RunProgressEventInput[];
  threshold: number;
  passAtKValues?: number[];
  generatedAt?: string;
}

export interface ProgressReportReference {
  run_id: string;
  threshold: number;
  best_score: number | null;
  milestones_reached: number;
  pass_at_k: PassAtKSummary[];
}

export function buildRunProgressReport(input: BuildRunProgressReportInput): RunProgressReport {
  const events = input.events.map(normalizeEvent).sort(eventSort);
  const start = startTime(events, input.generatedAt);
  const scored = events.filter((event) => scoreOf(event) !== null);
  const milestones = milestoneMap(events, scored, input.threshold, start);
  return parseRunProgressReport({
    schema_version: 1,
    run_id: input.runId,
    generated_at: input.generatedAt ?? new Date().toISOString(),
    threshold: input.threshold,
    progress_points: progressPoints(scored, milestones, start),
    milestones: PROGRESS_MILESTONE_NAMES.map((name) => milestones[name]),
    pass_at_k: passAtK(scored, input.threshold, input.passAtKValues ?? [1, 5, 10]),
    branch_lineage: branchLineage(events),
  });
}

export function parseRunProgressReport(value: unknown): RunProgressReport {
  const report = record(value, "progress report");
  exact(report, ["schema_version", "run_id", "generated_at", "threshold", "progress_points", "milestones", "pass_at_k", "branch_lineage"]);
  if (report.schema_version !== 1) throw new Error("schema_version must be 1");
  return {
    schema_version: 1,
    run_id: string(report.run_id, "run_id"),
    generated_at: string(report.generated_at, "generated_at"),
    threshold: number(report.threshold, "threshold"),
    progress_points: array(report.progress_points, "progress_points").map(parseProgressPoint),
    milestones: array(report.milestones, "milestones").map(parseMilestone),
    pass_at_k: array(report.pass_at_k, "pass_at_k").map(parsePassAtK),
    branch_lineage: array(report.branch_lineage, "branch_lineage").map(parseLineageEdge),
  };
}

export function progressReportReference(report: RunProgressReport): ProgressReportReference {
  return {
    run_id: report.run_id,
    threshold: report.threshold,
    best_score: report.progress_points.reduce<number | null>(
      (best, point) => (best === null || point.best_score > best ? point.best_score : best),
      null,
    ),
    milestones_reached: report.milestones.filter((milestone) => milestone.reached).length,
    pass_at_k: report.pass_at_k.map((summary) => ({ ...summary })),
  };
}

function normalizeEvent(event: RunProgressEventInput): RunProgressEvent {
  if (!("payload" in event) || !event.payload) return event as RunProgressEvent;
  const payload = event.payload;
  return {
    event_id: maybeString((event as RunProgressEvent).event_id) ?? (event.seq !== undefined ? `seq-${event.seq}` : ""),
    event_type: event.event ?? "",
    timestamp: event.ts ?? (event as RunProgressEvent).timestamp ?? "",
    generation_index: maybeNumber(payload.generation_index) ?? maybeNumber(payload.generation),
    hypothesis_node_id:
      maybeString(payload.hypothesis_node_id) ?? maybeString(payload.node_id) ?? maybeString(payload.child_id),
    parent_hypothesis_node_id: maybeString(payload.parent_hypothesis_node_id) ?? maybeString(payload.parent_id),
    candidate_id: maybeString(payload.candidate_id),
    score: maybeNumber(payload.score) ?? maybeNumber(payload.best_score),
    verifier_passed: maybeBoolean(payload.verifier_passed),
    gate_decision: maybeString(payload.gate_decision) ?? maybeString(payload.decision),
  };
}

function eventSort(a: RunProgressEvent, b: RunProgressEvent): number {
  return a.timestamp.localeCompare(b.timestamp) || a.event_id.localeCompare(b.event_id);
}

function startTime(events: RunProgressEvent[], generatedAt?: string): number {
  const timestamps = events.map((event) => Date.parse(event.timestamp)).filter(Number.isFinite);
  if (timestamps.length > 0) return Math.min(...timestamps);
  return generatedAt ? Date.parse(generatedAt) : Date.now();
}

function milestoneMap(
  events: RunProgressEvent[],
  scored: RunProgressEvent[],
  threshold: number,
  start: number,
): Record<ProgressMilestoneName, MilestoneTiming> {
  return {
    first_valid_candidate: milestone("first_valid_candidate", scored[0], start),
    first_passing_verifier: milestone("first_passing_verifier", scored.find((event) => event.verifier_passed), start),
    first_advancement: milestone(
      "first_advancement",
      events.find((event) => (event.gate_decision ?? event.decision) === "advance"),
      start,
    ),
    threshold_success: milestone(
      "threshold_success",
      scored.find((event) => (scoreOf(event) ?? 0) >= threshold),
      start,
    ),
  };
}

function milestone(name: ProgressMilestoneName, event: RunProgressEvent | undefined, start: number): MilestoneTiming {
  if (!event) return { name, reached: false, event_id: null, timestamp: null, elapsed_seconds: null, generation_index: null, hypothesis_node_id: null, score: null };
  return {
    name,
    reached: true,
    event_id: event.event_id,
    timestamp: event.timestamp,
    elapsed_seconds: elapsedSeconds(event.timestamp, start),
    generation_index: event.generation_index ?? null,
    hypothesis_node_id: event.hypothesis_node_id ?? null,
    score: scoreOf(event),
  };
}

function progressPoints(
  scored: RunProgressEvent[],
  milestones: Record<ProgressMilestoneName, MilestoneTiming>,
  start: number,
): ProgressPoint[] {
  const pointMilestones = new Map<string, ProgressMilestoneName[]>();
  for (const item of Object.values(milestones)) {
    if (!item.event_id) continue;
    pointMilestones.set(item.event_id, [...(pointMilestones.get(item.event_id) ?? []), item.name]);
  }

  let best = Number.NEGATIVE_INFINITY;
  return scored.map((event) => {
    const score = scoreOf(event) ?? 0;
    const improved = score > best;
    if (improved) best = score;
    return {
      event_id: event.event_id,
      timestamp: event.timestamp,
      elapsed_seconds: elapsedSeconds(event.timestamp, start),
      generation_index: event.generation_index ?? null,
      hypothesis_node_id: event.hypothesis_node_id ?? null,
      candidate_id: event.candidate_id ?? null,
      score,
      best_score: best,
      improved,
      milestone_names: pointMilestones.get(event.event_id) ?? [],
    };
  });
}

function passAtK(scored: RunProgressEvent[], threshold: number, values: number[]): PassAtKSummary[] {
  const scores = scored.map(scoreOf).filter((score): score is number => score !== null);
  return values.filter((k) => k > 0).map((k) => {
    const window = scores.slice(0, k);
    const successes = window.filter((score) => score >= threshold).length;
    return { k, trials_considered: window.length, successes, passed: successes > 0, best_score: window.length ? Math.max(...window) : null, threshold };
  });
}

function branchLineage(events: RunProgressEvent[]): BranchLineageEdge[] {
  const seen = new Set<string>();
  const edges: BranchLineageEdge[] = [];
  for (const event of events) {
    const parent = event.parent_hypothesis_node_id;
    const child = event.hypothesis_node_id;
    if (!parent || !child || parent === child) continue;
    const key = `${parent}\u0000${child}`;
    if (seen.has(key)) continue;
    seen.add(key);
    edges.push({ parent_hypothesis_node_id: parent, child_hypothesis_node_id: child, event_id: event.event_id, generation_index: event.generation_index ?? null });
  }
  return edges;
}

function parseProgressPoint(value: unknown): ProgressPoint {
  const item = record(value, "progress point");
  exact(item, ["event_id", "timestamp", "elapsed_seconds", "generation_index", "hypothesis_node_id", "candidate_id", "score", "best_score", "improved", "milestone_names"]);
  return {
    event_id: string(item.event_id, "event_id"),
    timestamp: string(item.timestamp, "timestamp"),
    elapsed_seconds: number(item.elapsed_seconds, "elapsed_seconds"),
    generation_index: nullableInteger(item.generation_index, "generation_index"),
    hypothesis_node_id: nullableString(item.hypothesis_node_id, "hypothesis_node_id"),
    candidate_id: nullableString(item.candidate_id, "candidate_id"),
    score: number(item.score, "score"),
    best_score: number(item.best_score, "best_score"),
    improved: boolean(item.improved, "improved"),
    milestone_names: array(item.milestone_names, "milestone_names").map(milestoneName),
  };
}

function parseMilestone(value: unknown): MilestoneTiming {
  const item = record(value, "milestone");
  exact(item, ["name", "reached", "event_id", "timestamp", "elapsed_seconds", "generation_index", "hypothesis_node_id", "score"]);
  return {
    name: milestoneName(item.name),
    reached: boolean(item.reached, "reached"),
    event_id: nullableString(item.event_id, "event_id"),
    timestamp: nullableString(item.timestamp, "timestamp"),
    elapsed_seconds: nullableNumber(item.elapsed_seconds, "elapsed_seconds"),
    generation_index: nullableInteger(item.generation_index, "generation_index"),
    hypothesis_node_id: nullableString(item.hypothesis_node_id, "hypothesis_node_id"),
    score: nullableNumber(item.score, "score"),
  };
}

function parsePassAtK(value: unknown): PassAtKSummary {
  const item = record(value, "pass_at_k");
  exact(item, ["k", "trials_considered", "successes", "passed", "best_score", "threshold"]);
  return {
    k: integer(item.k, "k"),
    trials_considered: integer(item.trials_considered, "trials_considered"),
    successes: integer(item.successes, "successes"),
    passed: boolean(item.passed, "passed"),
    best_score: nullableNumber(item.best_score, "best_score"),
    threshold: number(item.threshold, "threshold"),
  };
}

function parseLineageEdge(value: unknown): BranchLineageEdge {
  const item = record(value, "branch lineage edge");
  exact(item, ["parent_hypothesis_node_id", "child_hypothesis_node_id", "event_id", "generation_index"]);
  return {
    parent_hypothesis_node_id: string(item.parent_hypothesis_node_id, "parent_hypothesis_node_id"),
    child_hypothesis_node_id: string(item.child_hypothesis_node_id, "child_hypothesis_node_id"),
    event_id: string(item.event_id, "event_id"),
    generation_index: nullableInteger(item.generation_index, "generation_index"),
  };
}

function scoreOf(event: RunProgressEvent): number | null {
  return typeof event.score === "number" && Number.isFinite(event.score) ? event.score : null;
}

function elapsedSeconds(timestamp: string, start: number): number {
  return (Date.parse(timestamp) - start) / 1000;
}

function milestoneName(value: unknown): ProgressMilestoneName {
  if (typeof value !== "string" || !PROGRESS_MILESTONE_NAMES.includes(value as ProgressMilestoneName)) throw new Error("unknown milestone name");
  return value as ProgressMilestoneName;
}

function record(value: unknown, label: string): Record<string, unknown> {
  if (!value || typeof value !== "object" || Array.isArray(value)) throw new Error(`${label} must be an object`);
  return value as Record<string, unknown>;
}

function exact(item: Record<string, unknown>, allowed: readonly string[]): void {
  const allowedSet = new Set(allowed);
  const extra = Object.keys(item).filter((key) => !allowedSet.has(key));
  if (extra.length) throw new Error(`unexpected field(s): ${extra.sort().join(", ")}`);
}

function string(value: unknown, label: string): string {
  if (typeof value !== "string" || value.length === 0) throw new Error(`${label} must be a string`);
  return value;
}

function nullableString(value: unknown, label: string): string | null {
  return value === null || value === undefined ? null : string(value, label);
}

function number(value: unknown, label: string): number {
  if (typeof value !== "number" || !Number.isFinite(value)) throw new Error(`${label} must be a number`);
  return value;
}

function nullableNumber(value: unknown, label: string): number | null {
  return value === null || value === undefined ? null : number(value, label);
}

function integer(value: unknown, label: string): number {
  if (typeof value !== "number" || !Number.isInteger(value)) throw new Error(`${label} must be an integer`);
  return value;
}

function nullableInteger(value: unknown, label: string): number | null {
  return value === null || value === undefined ? null : integer(value, label);
}

function boolean(value: unknown, label: string): boolean {
  if (typeof value !== "boolean") throw new Error(`${label} must be boolean`);
  return value;
}

function array(value: unknown, label: string): unknown[] {
  if (!Array.isArray(value)) throw new Error(`${label} must be an array`);
  return value;
}

function maybeString(value: unknown): string | undefined {
  return typeof value === "string" && value.length ? value : undefined;
}

function maybeNumber(value: unknown): number | undefined {
  return typeof value === "number" && Number.isFinite(value) ? value : undefined;
}

function maybeBoolean(value: unknown): boolean | undefined {
  return typeof value === "boolean" ? value : undefined;
}
